### PR TITLE
TT-10564 Splunk backoff retry

### PR DESCRIPTION
**Pull request by mativm02**
This PR is a duplicated of https://github.com/TykTechnologies/tyk-pump/pull/745

Introduce an expotential backoff retry mechanism for sending logs to spunk to mitigate the chance of losing logs if pump fails to send logs to spunk (i.e. due to network issue or rate limit being hit). Also fixes bug where pump not closing connections after receiving response. Also spunk responses were not being checked i.e. a `404` could be returned and pump would still mark as successful pump export.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [x] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: success** :white_check_mark: 
Branch used: refs/pull/758/merge
Commit:  
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk-pump/actions/runs/7350075274)

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/758/merge
Commit:  
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk-pump/actions/runs/7350075274)

**Comment by buger** [comment]
**API tests result - postgres15-murmur64 env: success** :white_check_mark: 
Branch used: refs/pull/758/merge
Commit:  
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk-pump/actions/runs/7350075274)

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/758/merge
Commit:  
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk-pump/actions/runs/7350075274)

File: README.md [hunk 1]
```diff
@@ -775,6 +775,7 @@ Setting up Splunk with a _HTTP Event Collector_
 - `ignore_tag_prefix_list`: (optional) Choose which tags to be ignored by the Splunk Pump. Keep in mind that the tag name and value are hyphenated. Type: Type: String Array `[] string`. Default value is `[]`
 - `enable_batch`: If this is set to `true`, pump is going to send the analytics records in batch to Splunk. Type: Boolean. Default value is `false`.
 - `max_content_length`: Max content length in bytes to be sent in batch requests. It should match the `max_content_length` configured in Splunk. If the purged analytics records size don't reach the amount of bytes, they're send anyways in each `purge_loop`. Type: Integer. Default value is 838860800 (~ 800 MB), the same default value as Splunk config.
+- `max_retries`: Max number of retries if failed to send requests to splunk HEC. Default value is `0`.
```
[the remark this answers is not in the document]
Comment (sredxny) [reply README.md:778]: also we should say when a retry is gonna happen (explain which status codes will be considered as errors)

File: pumps/splunk_test.go [hunk 2]
```diff
@@ -79,6 +93,44 @@ func TestSplunkInit(t *testing.T) {
...
+	server.Config.SetKeepAlivesEnabled(false)
+	server.Start()
+
+	defer server.Close()
+
+	pmp := SplunkPump{}
+
+	cfg := make(map[string]interface{})
+	cfg["collector_token"] = testToken
+	cfg["max_retries"] = 1
```
Comment (sredxny) [inline pumps/splunk_test.go:108]: is possible to add more use-cases? Eg: max_retries: {-1, 0, 1, 3}

**Review by sredxny — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mativm02) [reply README.md]: comment improved. thanks!

**Review by sredxny — APPROVED** [review]

**Comment by mativm02** [comment]
It's merged! Thank you for your contribution @domsolutions, it's really appreciated.

**Comment by domsolutions** [comment]
> It's merged! Thank you for your contribution @domsolutions, it's really appreciated.

No problem!

**Pull request merged**